src/masternodeman.cpp

```diff
@@ -581,99 +581,88 @@ masternode_info_t CMasternodeMan::FindRandomNotInVec(const std::vector<COutPoint
     return masternode_info_t();
 }
 
-CMasternodeMan::score_pair_m_t CMasternodeMan::GetMasternodeScores(int nBlockHeight, uint256 blockHash, int nMinProtocol)
+bool CMasternodeMan::GetMasternodeScores(CMasternodeMan::score_pair_vec_t& vecMasternodeScoresRet, int nBlockHeight, int nMinProtocol)
 {
-    score_pair_m_t vecMasternodeScores;
+    vecMasternodeScoresRet.clear();
 
-    if(!masternodeSync.IsMasternodeListSynced())
-        return vecMasternodeScores;
+    if (!masternodeSync.IsMasternodeListSynced())
+        return false;
 
-    assert(blockHash != uint256());
+    // make sure we know about this block
+    uint256 blockHash = uint256();
+    if (!GetBlockHash(blockHash, nBlockHeight)) {
+        LogPrintf("CMasternodeMan::%s -- ERROR: GetBlockHash() failed at nBlockHeight %d\n", __func__, nBlockHeight);
+        return false;
+    }
 
-    // lock is required and should be acquired outside
-    AssertLockHeld(cs);
+    LOCK(cs);
```

Comment:
Pulling the locking into this method (instead of forcing the caller to hold it), creates a race condition. This method returns MNs by pointer and not by value. As the caller does not hold the lock anymore, it's not guaranteed that the pointers are still valid.

Reply:
hah, I knew there was a reason why I did this in the first version :D